v3iofs/fs.py

```diff
@@ -62,43 +62,53 @@ def ls(self, path, detail=True, **kwargs):
             path=path,
             get_all_attributes=True,
             raise_for_status=v3io.dataplane.RaiseForStatus.never,
+            marker=marker,
         )
 
         # Ignore 404's here
         if resp.status_code not in {200, 404}:
             raise Exception(
                 f'{resp.status_code} received while accessing {path!r}')
 
-        # If not data, try to find file in parent directory
-        if not _has_data(resp):
-            return [self._ls_file(container, path, detail)]
-
         out = (
                 _resp_dirs(resp, container, detail) +
                 _resp_files(resp, container, detail)
         )
 
-        if not out:
-            raise FileNotFoundError(f'{full_path!r} not found')
+        if not marker:
+            # first time in
+            if not _has_data(resp):
+                return [self._ls_file(container, path, detail)]
+            if not out:
+                raise FileNotFoundError(f'{full_path!r} not found')
+
+        if hasattr(resp.output, 'next_marker') and resp.output.next_marker:
+            marker = resp.output.next_marker
+            out = out + self.ls(path, detail, marker)
```

Comment:
`path` here is not the original argument, because the container part has been removed. So this recursive call fails silently and only the first page is ever returned.

Ideally we would want to:
1) Pass the original path to fix this issue
2) Fix the issue of silent failure
3) Add a test

Reply:
Maybe also better to drop the recursion, because it introduces some confusiong (_ls_file isn't really supposed to be called inside a recursive ls call, but that's only clear from knowing the system, not from the code).